chain-adapter/src/main/kotlin/com/d3/chainadapter/adapter/ChainAdapter.kt

```diff
@@ -0,0 +1,150 @@
+    }
+
+    /**
+     * Initiates Iroha chain listener logic
+     * @param channel - channel that is used to publish Iroha blocks
+     * @param onIrohaListenError - function that will be called on Iroha chain listener error
+     */
+    private fun initIrohaChainListener(channel: Channel, onIrohaListenError: () -> Unit) {
+        irohaChainListener.getBlockObservable()
+            .map { observable ->
+                observable.subscribeOn(Schedulers.from(subscriberExecutorService))
+                    .subscribe({ block ->
+                        publishUnreadLatch.await()
+                        // Send only not read Iroha blocks
+                        if (block.blockV1.payload.height > lastReadBlock.get()) {
+                            onNewBlock(channel, block)
+                        }
+                    }, { ex ->
+                        logger.error("Error on Iroha chain listener occurred", ex)
+                        onIrohaListenError()
+                    })
+            }
+    }
+
+    /**
+     * Publishes unread blocks
+     * @param channel - RabbitMQ channel that is used to publish Iroha blocks
+     */
+    private fun publishUnreadIrohaBlocks(channel: Channel) {
+        var lastProcessedBlock = lastReadBlockProvider.getLastBlock()
+        while (true) {
+            lastProcessedBlock++
+            logger.info { "Try read Iroha block $lastProcessedBlock" }
+            val response = getBlockRawResponse(queryAPI, lastProcessedBlock)
+            if (response.hasErrorResponse()) {
+                val errorResponse = response.errorResponse
+                if (isNoMoreBlocksError(errorResponse)) {
+                    logger.info { "Done publishing unread blocks" }
+                    break
+                } else {
+                    throw Exception("Cannot get block. ${getErrorMessage(errorResponse)}")
+                }
+            } else if (response.hasBlockResponse()) {
+                onNewBlock(channel, response.blockResponse!!.block)
+            }
+        }
+        publishUnreadLatch.countDown()
+    }
+
+    /**
+     * Checks if no more blocks
+     * @param errorResponse - error response to check
+     * @return true if no more blocks to read
+     */
+    private fun isNoMoreBlocksError(errorResponse: QryResponses.ErrorResponse) =
+        errorResponse.errorCode == BAD_IROHA_BLOCK_HEIGHT_ERROR_CODE
+
+    /**
+     * Publishes new block to RabbitMQ
+     * @param channel - channel that is used to publish blocks
+     * @param block - Iroha block to publish
+     */
+    private fun onNewBlock(channel: Channel, block: BlockOuterClass.Block) {
+        val message = block.toByteArray()
+        channel.basicPublish(
+            rmqConfig.irohaExchange,
+            "",
+            MessageProperties.MINIMAL_PERSISTENT_BASIC,
+            message
+        )
+        logger.info { "Block pushed" }
+        // Save last read block
+        lastReadBlockProvider.saveLastBlock(block)
```

Comment:
do we need to use `block` as a parameter? What are the other possible implementations of such provider?